### PR TITLE
bug fixed for API to get user's repos

### LINKED ISSUE #1540: Panic when listing user repos using API and not authenticated
- Gitea version (or commit ref): 1.1.0+138-gc58708d3
- Git version: 2.7.4
- Operating system: Ubuntu 16.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/api/v1/users/cez81/repos
  - [ ] No
  - [ ] Not relevant
- Log gist:
```
/home/jonas/go/src/code.gitea.io/gitea/routers/api/v1/user/repo.go:12 (0x101b389)
	listUserRepos: showPrivateRepos := (ctx.User.ID == userID || ctx.User.IsAdmin) && ctx.IsSigned
/home/jonas/go/src/code.gitea.io/gitea/routers/api/v1/user/repo.go:40 (0x101b764)
	ListUserRepos: listUserRepos(ctx, user)
```

## Description
Trying to get users repositories using the API when not signed in will cause panic on line https://github.com/go-gitea/gitea/blob/master/routers/api/v1/user/repo.go#L12.
ctx.User is not populated unless signed in.

Github seems to allow this operation even when nog logged in so I guess that should be correct behaviour?

### PR DESCRIPTION
**Pull request by lunny**
As title. Will fix #1540. And how to test APIs? Maybe we need a framework like intergration test to do that?

### CODE REVIEW

**Comment by sapk** [comment]
LGTM

File: routers/api/v1/user/repo.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 // listUserRepos - List the repositories owned by the given user.
 func listUserRepos(ctx *context.APIContext, u *models.User) {
 	userID := u.ID
-	showPrivateRepos := (ctx.User.ID == userID || ctx.User.IsAdmin) && ctx.IsSigned
+	showPrivateRepos := ctx.IsSigned && (ctx.User.ID == userID || ctx.User.IsAdmin)
```
Comment (cez81) [inline routers/api/v1/user/repo.go:12]: ctx.User is also used in `getAccessibleRepos`called from https://github.com/lunny/gitea/blob/07ad1e27dd8fbebb90b53590a958115e7c680130/routers/api/v1/user/repo.go#L18

**Comment by sapk** [comment]
For testing you can make integration test and request the API directly ?

**Comment by bkcsoft** [comment]
@lunny could you fix #1539 at the same time? (same bug really)

**Comment by lunny** [comment]
@sapk I have added a test
@bkcsoft I have sent #1624 to fix #1539

**Comment by lunny** [comment]
@bkcsoft oh. It's not. I will send another one to fix #1539

**Comment by sapk** [comment]
@lunny Perfect. Still LGTM for me ^^

**Comment by bkcsoft** [comment]
LGTM

File: integrations/api_repo_test.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+	"code.gitea.io/gitea/models"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAPIUserReposNotLogin(t *testing.T) {
+	assert.NoError(t, models.LoadFixtures())
+
+	req, err := http.NewRequest("GET", "/api/v1/users/user1/repos", nil)
```
Comment (sapk) [inline integrations/api_repo_test.go]: user1 don't have repos in mock value so maybe we could also test user2 repo listing?

https://github.com/go-gitea/gitea/blob/master/models/fixtures/repository.yml

**Comment by lunny** [comment]
waiting for #1627

**Comment by lunny** [comment]
let L-G-T-M work

**Pull request merged**